webhook/resources/engine/mutator.go

```diff
@@ -53,5 +56,25 @@ func mutateEngine(newObj runtime.Object) (admission.PatchOps, error) {
 		patchOps = append(patchOps, `{"op": "replace", "path": "/spec/upgradedReplicaAddressMap", "value": {}}`)
 	}
 
+	labels := engine.Labels
+	if labels == nil {
+		labels = map[string]string{}
+	}
+	patchOp, err := common.GetLonghornLabelsPatchOp(engine, labels)
+	if err != nil {
+		err := errors.Wrapf(err, "failed to get label patch for engine %v", engine.Name)
+		return nil, werror.NewInvalidError(err.Error(), "")
+	}
+	patchOps = append(patchOps, patchOp)
+
+	if needFinalizer {
```

Comment:
The update function does not need to add finalizers is because when a volume is deleted, both replica and engine finalizers are removed in the API first.

In `GetLonghornFinalizerPatchOp`, the finalizer was added back, which caused the replica and engine to not be removed properly.

https://github.com/longhorn/longhorn-manager/blob/d05186d9741bd98c1bf86c0cc51ae370979b899a/datastore/longhorn.go#L1158-L1175

Reply:
You are right :+1: !